server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/services/BuildServerService.java

```diff
@@ -161,31 +164,51 @@ public void onBuildExit() {
 
   public Either<ResponseError, SourcesResult> buildTargetSources(SourcesParams sourcesParams) {
     LOGGER.info("buildTargetSources call with param: {}", sourcesParams);
+    try {
+      TargetRulesResolver<SourcesItem> targetRulesResolver =
+          TargetRulesResolver.withBazelRunnerAndMapper(
+              bazelRunner, this::mapBuildRuleToSourcesItem);
 
-    TargetRulesResolver<SourcesItem> targetRulesResolver =
-        TargetRulesResolver.withBazelRunnerAndMapper(bazelRunner, this::mapBuildRuleToSourcesItem);
-
-    List<SourcesItem> sourceItems =
-        targetRulesResolver.getItemsForTargets(sourcesParams.getTargets());
+      List<SourcesItem> sourceItems =
+          targetRulesResolver.getItemsForTargets(sourcesParams.getTargets());
 
-    SourcesResult sourcesResult = new SourcesResult(sourceItems);
+      SourcesResult sourcesResult = new SourcesResult(sourceItems);
 
-    return Either.forRight(sourcesResult);
+      return Either.forRight(sourcesResult);
+    } catch (Exception e) {
+      ResponseError fail =
+          new ResponseError(ResponseErrorCode.InternalError, e.getMessage(), e.getStackTrace());
+      return Either.forLeft(fail);
+    }
   }
 
   private SourcesItem mapBuildRuleToSourcesItem(Build.Rule rule) {
     BuildTargetIdentifier ruleLabel = new BuildTargetIdentifier(rule.getName());
     List<SourceItem> items = serverBuildManager.getSourceItems(rule, ruleLabel);
-    List<String> roots = getRuleRoots(rule);
+    List<String> roots = getRuleRoots(items);
 
     return createSourcesForLabelAndItemsAndRoots(ruleLabel, items, roots);
   }
 
-  private List<String> getRuleRoots(Build.Rule rule) {
-    String sourcesRootUriString = serverBuildManager.getSourcesRoot(rule.getName());
-    Uri uri = Uri.fromAbsolutePath(sourcesRootUriString);
-
-    return ImmutableList.of(uri.toString());
+  private List<String> getRuleRoots(List<SourceItem> items) {
+    Set<String> sourceRootUris =
+        items.stream()
+            .map(item -> item.getUri())
+            .map(
+                uri -> {
+                  try {
+                    URL url = new URL(uri);
+                    return url.toURI();
+                  } catch (Exception e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .map(SourceRootGuesser::getSourcesRoot)
+            .map(Uri::fromAbsolutePath)
+            .map(uri -> uri.toString())
+            .collect(Collectors.toSet());
+
+    return ImmutableList.copyOf(sourceRootUris);
```

Comment:
could be replaced by `.distinct()` (look previous version of review)

Reply:
Done